File: haskell/2021/day_02.hs
run :: Num a => Bool -> [(String, a)] -> (a, a, a) -> a
run _ [] (_, x, y) = x * y
run adv ((dir, val):cmds) (aim, x, y) = run adv cmds $ case (adv, dir) of
    (False, "forward") -> (aim, x + val, y)
    (True, "forward") -> (aim, x + val, y + aim * val)
    (False, "down") -> (aim, x, y + val)
    (True, "down") -> (aim + val, x, y)
    (False, "up") -> (aim, x, y - val)
    (True, "up") -> (aim - val, x, y)
    _ -> error "Invalid input"

main :: IO ()
main = do
    text <- readFile "../../input/2021/input_02.txt"
    let cmds = [(s, read v :: Int) | s:v:_ <- map words $ lines text]
    putStrLn $ "Part 1: " ++ show (run False cmds (0, 0, 0))
    putStrLn $ "Part 2: " ++ show (run True cmds (0, 0, 0))

File: rust/2021/src/bin/day_02.rs
use std::fs;

fn run(cmds: &[(&str, u64)], advanced: bool) -> u64 {
    let (mut aim, mut x, mut y) = (0, 0, 0);
    for (dir, val) in cmds {
        match *dir {
            "forward" => {
                x += val;
                if advanced {
                    y += aim * val
                }
            }
            "down" => {
                if advanced {
                    aim += val
                } else {
                    y += val
                }
            }
            "up" => {
                if advanced {
                    aim -= val
                } else {
                    y -= val
                }
            }
            _ => unreachable!(),
        }
    }
    x * y
}

fn main() {
    let input = fs::read_to_string("../input/2021/input_02.txt").unwrap();
    let cmds: Vec<_> = input
        .lines()
        .map(|line| {
            let (a, b) = line.split_once(' ').unwrap();
            (a, b.parse::<u64>().unwrap())
        })
        .collect();
    println!("Part 1: {}", run(&cmds, false));
    println!("Part 2: {}", run(&cmds, true));
}
